File: python/ach.py
#!/usr/bin/env python

## Copyright (c) 2013, Georgia Tech Research Corporation
## All rights reserved.
##
## Author(s): Neil T. Dantam <ntd@gatech.edu>
## Georgia Tech Humanoid Robotics Lab
## Under Direction of Prof. Mike Stilman <mstilman@cc.gatech.edu>
##
## Redistribution and use in source and binary forms, with or without
## modification, are permitted provided that the following conditions
## are met:
##
##     * Redistributions of source code must retain the above
##       copyright notice, this list of conditions and the following
##       disclaimer.
##     * Redistributions in binary form must reproduce the above
##       copyright notice, this list of conditions and the following
##       disclaimer in the documentation and/or other materials
##       provided with the distribution.
##
## THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
## ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
## LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
## FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
## COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
## INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
## (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
## SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
## HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
## STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
## OF THE POSSIBILITY OF SUCH DAMAGE.

## FILE: ach.py
## DESC: python module to access ach

import ach_py

from ach_py import \
    ACH_OK,\
    ACH_OVERFLOW,\
    ACH_INVALID_NAME,\
    ACH_BAD_SHM_FILE,\
    ACH_FAILED_SYSCALL,\
    ACH_STALE_FRAMES,\
    ACH_MISSED_FRAME,\
    ACH_TIMEOUT,\
    ACH_EEXIST,\
    ACH_ENOENT,\
    ACH_CLOSED,\
    ACH_BUG,\
    ACH_EINVAL,\
    ACH_CORRUPT,\
    ACH_BAD_HEADER,\
    ACH_EACCES,\
    ACH_CANCELED,\
    ACH_O_WAIT,\
    ACH_O_LAST, \
    AchException, \
    ACH_DEFAULT_FRAME_COUNT, \
    ACH_DEFAULT_FRAME_SIZE

class Channel:
    """ An Ach channel."""

    def __init__(self, name=None, frame_count=ACH_DEFAULT_FRAME_COUNT, frame_size=ACH_DEFAULT_FRAME_SIZE):
        '''Construct a channel object.

        name -- if provided, the name of the channel to open or create
        '''
        self.pointer = None
        self.name = None
        if name:
            self.open(name, frame_count, frame_size)

    def __del__(self):
        '''Close the channel.'''
        if ach_py and self.pointer:
            self.close()

    def open(self, name, frame_count=ACH_DEFAULT_FRAME_COUNT, frame_size=ACH_DEFAULT_FRAME_SIZE):
        '''Open channel with the given name.
        If channel does not exist, create it.'''
        assert( not self.pointer )
        self.pointer = ach_py.open_channel(name, frame_count, frame_size)
        self.name = name

    def close(self):
        '''Close the channel.'''
        assert(self.pointer)
        ach_py.close_channel(self.pointer)
        self.pointer = None
        self.name = None

    def put(self, buf):
        '''Put a buffer into the channel.

        buf -- an object providing the buffer interface'''
        assert(self.pointer)
        ach_py.put_buf( self.pointer, buf )

    def get(self, buf, wait=False, last=False):
        '''Get a message from channel and write to buffer.

        buf -- an object providing the buffer interface
        wait -- wait until new message posts
        last -- get most recent message

        returns a tuple of (ach_status, frame_size)'''
        assert(self.pointer)
        return ach_py.get_buf( self.pointer, buf, wait, last )

    def result_string( self, status ):
        '''Get a string description of integer code status.'''
        return ach_py.result_string(status)

    def flush( self ):
        '''Flush unseen messages from the channel.'''
        assert(self.pointer)
        ach_py.flush_channel(self.pointer)

    def chmod( self, mode ):
        '''Set channel permissions to mode.'''
        assert(self.pointer)
        ach_py.chmod_channel(self.pointer, mode)

    def unlink( self ):
        '''Unlink (delete) the channel's underlying file.
        Note: this does not close the channel in this or any other process.'''
        assert(self.pointer)
        ach_py.unlink_channel(self.name)

File: src/ach.c
/* -*- mode: C; c-basic-offset: 4 -*- */
/* ex: set shiftwidth=4 tabstop=4 expandtab: */
/*
 * Copyright (c) 2011-2013, Georgia Tech Research Corporation
 * All rights reserved.
 *
 * Author(s): Neil T. Dantam <ntd@gatech.edu>
 * Georgia Tech Humanoid Robotics Lab
 * Under Direction of Prof. Mike Stilman <mstilman@cc.gatech.edu>
 *
 *
 * This file is provided under the following "BSD-style" License:
 *
 *
 *   Redistribution and use in source and binary forms, with or
 *   without modification, are permitted provided that the following
 *   conditions are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *
 *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *   POSSIBILITY OF SUCH DAMAGE.
 *
 */

#ifdef HAVE_CONFIG
#include "config.h"
#endif //HAVE_CONFIG

#include <ipcbench.h>
#include <ach.h>

static ach_channel_t *channel;
static size_t cnt;
static struct pollfd *pfd;

#define CHAN_NAME "ipcbench"

static void s_init(enum ach_map map, size_t channel_cnt) {
    cnt = channel_cnt;

    channel = (struct ach_channel*)calloc(cnt, sizeof(struct ach_channel));

    for( size_t i = 0; i < cnt; i ++ ) {
        char buf[64];
        snprintf( buf, sizeof(buf), CHAN_NAME "-%d", i );

        ach_status_t r = ach_unlink(buf);               /* delete first */

        if( !( ACH_OK == r || ACH_ENOENT == r) ) {
            fprintf(stderr, "ach_unlink: %s\n", ach_result_to_string(r) );
            abort();
        }

        struct ach_create_attr attr;
        ach_create_attr_init(&attr);
        if( ACH_OK != (r = ach_create_attr_set_map(&attr,map)) ) {
            fprintf(stderr, "ach_create_attr_set_map: %s\n", ach_result_to_string(r));
            abort();
        }

        if( ACH_OK != (r = ach_create(buf, 10,
                                      sizeof(struct timespec), &attr)) ) {
            fprintf(stderr, "ach_create: %s\n", ach_result_to_string(r) );
            abort();
        }


        for(;;) {
            r = ach_open( &channel[i], buf, NULL );
            if( ACH_OK == r ) {
                break;
            } else if( ach_status_match(r, ACH_MASK_ENOENT | ACH_MASK_EACCES) ) {
                usleep(1000);     /* Race against udev */
            } else {
                fprintf(stderr, "ach_open", ach_result_to_string(r) );
                abort();
            }
        }
    }

    if( cnt > 1 ) {
        assert( ACH_MAP_KERNEL == map );
        pfd = (struct pollfd*)calloc(cnt, sizeof(struct pollfd));
        for( size_t i = 0; i < cnt; i ++ ) {
            enum ach_status r = ach_channel_fd(&channel[i], &pfd[i].fd);
            if( ACH_OK != r ) {
                fprintf(stderr, "ach_channel_fd: %s\n", ach_result_to_string(r));
                abort();
            }
            pfd[i].events = POLLIN;
        }
    }
}


static void s_init_user(size_t n)
{
    s_init(ACH_MAP_USER, n);
}

static void s_init_kernel(size_t n)
{
    s_init(ACH_MAP_KERNEL, n);
}

static void s_destroy(void) {
    for( size_t i = 0; i < cnt; i ++ ) {
        char buf[64];
        snprintf( buf, sizeof(buf), CHAN_NAME "-%d", i );

        enum ach_status r = ach_unlink(buf);
        if( ACH_OK != r ) {
            fprintf(stderr, "ach_unlink: %s\n", ach_result_to_string(r));
            abort();
        }
    }
}

static void s_send( const struct timespec *ts ) {
    size_t i = pubnext(cnt);
    ach_status_t r = ach_put( &channel[i], ts, sizeof(*ts) );
    if( ACH_OK != r ) {
        fprintf(stderr, "ach_put: %s\n", ach_result_to_string(r) );
        abort();
    }
}

static void s_recv( struct timespec *ts ) {
    size_t i = pollin( pfd, cnt );
    size_t fs;
    ach_status_t r = ach_get( &channel[i], ts, sizeof(*ts),
                              &fs, NULL, ACH_O_WAIT | ACH_O_LAST );
    if( ! (ACH_OK==r || ACH_MISSED_FRAME == r) ) {
        fprintf(stderr, "ach_get: %s\n", ach_result_to_string(r) );
        abort();
    }
    if( sizeof(*ts) != fs ) {
        abort();
    }
}

struct ipcbench_vtab ipc_bench_vtab_ach_user = {
    .init = s_init_user,
    .send = s_send,
    .recv = s_recv,
    .destroy = s_destroy,
};

struct ipcbench_vtab ipc_bench_vtab_ach_kernel = {
    .init = s_init_kernel,
    .send = s_send,
    .recv = s_recv,
    .destroy = s_destroy,
};
